kdbx/src/main/java/org/linguafranca/pwdb/kdbx/KdbxKeyFile.java

```diff
@@ -45,33 +52,61 @@ public class KdbxKeyFile {
      * @return the key
      */
     public static byte[] load(InputStream inputStream) {
+
         try {
-            DocumentBuilder documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
-            Document doc = documentBuilder.parse(inputStream);
-            String version = (String) xpath.evaluate("//KeyFile/Meta/Version/text()", doc, XPathConstants.STRING);
-            String data = (String) xpath.evaluate("//KeyFile/Key/Data/text()", doc, XPathConstants.STRING);
-            if (data == null) {
-                return null;
-            }
-            if (version.equals("2.0")) {
-               
-                byte[] hexData = Hex.decodeHex(data.replaceAll("\\s",""));
-               
-                MessageDigest md = Encryption.getSha256MessageDigestInstance();
-                byte[] computedHash = md.digest(hexData);
-               
-                String hashToCheck = (String) xpath.evaluate("//KeyFile/Key/Data/@Hash", doc, XPathConstants.STRING);
-                byte[] verifiedHash = Hex.decodeHex(hashToCheck);
-                
-                for(int i = 0; i < verifiedHash.length; i++) {
-                    if(computedHash[i] != verifiedHash[i]) {
+            byte[] inputBytes = ByteStreams.toByteArray(inputStream);
```

Comment:
in the case where we are taking a digest of an arbitrary file, it may be megabytes long, and we would not want to read it all in to memory as a byte array. 

One way of doing this would be to wrap the InputStream in a PushbackInputStream - try to read 65 bytes from the stream, if you end up with 32 or 64 then you have your buffer otherwise push the buffer back and use the stream for XML or for doing a digest of the arbitrary file?

Reply:
Sounds good!